### PR TITLE
Fix IRQ-related socket locks

### PR DESCRIPTION
**Pull request by lrh2000**
This fixes **"Bug 3: Dead lock: Missing `irq_disabled` when acquiring various `StreamSocket`-related locks"** mentioned in https://github.com/asterinas/asterinas/issues/1406#issuecomment-2392759804.

Currently, we hold too many locks in IRQ handlers, which means that all involved locks have to be protected with IRQ-disabled guards, which seems a bit excessive:
 https://github.com/asterinas/asterinas/blob/96efd620072a0cbdccc95b58901894111f17bb3a/kernel/src/net/socket/ip/stream/mod.rs#L640-L648

This PR tries to make some more changes by delaying the state transition (lines 642 - 648 above) until some user events occur.
- Note that `IoEvents::OUT` is added by `ConnectingStream::update_io_events` when the state transition is needed, but the actual state transition is delayed until one of `StreamSocket::update_options_and_state`/`read_updated_state`/`write_updated_state` is called.
- I think things can be better with https://github.com/asterinas/asterinas/issues/1356, so instead of letting `ConnectingStream::update_io_events` add `IoEvents::OUT`, we can let the target state calculate the I/O events in `StreamSocket::poll`.

### CODE REVIEW

File: kernel/src/net/socket/ip/stream/mod.rs [hunk 1]
```diff
@@ -570,14 +608,7 @@ impl Socket for StreamSocket {
...
             tcp_maxseg: MaxSegment => {
-                // It will always return the default MSS value defined above for an unconnected socket
-                // and always return the actual current MSS for a connected one.
-
-                // FIXME: how to get the current MSS?
-                let maxseg = match self.state.read().as_ref() {
-                    State::Init(_) | State::Listen(_) | State::Connecting(_) => DEFAULT_MAXSEG,
-                    State::Connected(_) => options.tcp.maxseg(),
-                };
+                let maxseg = options.tcp.maxseg();
```
Comment (StevenJiang1110) [inline kernel/src/net/socket/ip/stream/mod.rs:611]: Why changing the maxseg option here? It seems unrelated to the bug?

**Review by StevenJiang1110 — COMMENTED** [review]
This PR looks fine to me! I only have minor comments!

File: kernel/src/net/socket/ip/stream/connecting.rs [hunk 2]
```diff
@@ -80,35 +77,39 @@ impl ConnectingStream {
...
+            pollee.add_events(IoEvents::OUT);
+
+            // Add `IoEvents::OUT` because the man pages say "EINPROGRESS [..] It is possible to
+            // select(2) or poll(2) for completion by selecting the socket for writing". For
+            // details, see <https://man7.org/linux/man-pages/man2/connect.2.html>.
+            //
+            // TODO: It is better to do the state transition and let `ConnectedStream` or
+            // `InitStream` set the correct I/O events. However, the state transition is delayed
+            // because we're probably in IRQ handlers. Maybe mark the `pollee` as obsolete and
+            // re-calculate the I/O events in `poll`.
```
[the remark this answers is not in the document]
Comment (lrh2000) [reply kernel/src/net/socket/ip/stream/connecting.rs:112]: Yes, but there are two occurrences of `polee.add_events(IoEvents::OUT)`. I just don't know which one to put the comments before.

File: kernel/src/net/socket/ip/stream/mod.rs (hunk 1 above)
Comment (lrh2000) [reply kernel/src/net/socket/ip/stream/mod.rs:611]: I have checked the [Linux implementation] (https://elixir.bootlin.com/linux/v6.11.2/source/net/ipv4/tcp.c#L4054). Nowhere does it implement the original comments:
```
                // It will always return the default MSS value defined above for an unconnected socket
                // and always return the actual current MSS for a connected one.
```
"It will always return the default MSS value defined above for an unconnected socket" This just looks like a false claim, and it is _not_ the Linux behavior. So I'm confused as to how this code is introduced.

The reason I change this is because I searched all `state.read()` and tried to replace it with `read_updated_state()`. However, I can't replace this with the MAXSEG option because if I replace it, there will be a chance of a deadlock (note that we have already held the write lock of `option` at this point).

Comment (StevenJiang1110) [reply kernel/src/net/socket/ip/stream/mod.rs:611]: > So I'm confused as to how this code is introduced.

I remember this code is introduced to pass a gvisor test, but I haven't checked the exact Linux behavior at that time.

But since the gvisor socket test is not added now, the change is acceptable for me since CI is not broken.

File: kernel/src/net/socket/ip/stream/connecting.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (StevenJiang1110) [reply kernel/src/net/socket/ip/stream/connecting.rs:112]: I used to think that these comments were meant to explain `pollee.add_events(IoEvents::OUT)` on line 103. Line 94 is quite intuitive; by definition, we should be adding OUT events for sockets that can send data, so additional explanation seems unnecessary.

If the comments are intended for both line 103 and line 94, it might be helpful to include an introductory line to clarify this, such as:

```rust
// The following comment applies to both line 94 and line 103:
```

[the remark this answers is not in the document]
Comment (lrh2000) [reply kernel/src/net/socket/ip/stream/connecting.rs:112]: > The following comment applies to both line 94 and line 103:

Having line numbers here won't work, they are too easy to break (e.g. inserting a line in the same file but not around the `update_io_events`).

> If the comments are intended for both line 103 and line 94, it might be helpful to include an introductory line to clarify this

Well, I think the wording in the man pages is clear enough, it is not specific to a successful `connect()` or a failed `connect()`. Moreover, I additionally wrote "let **`ConnectedStream` or `InitStream`** set the correct I/O events".

File: kernel/src/net/socket/ip/stream/mod.rs (hunk 1 above)
Comment (lrh2000) [reply kernel/src/net/socket/ip/stream/mod.rs:611]: > But since the gvisor socket test is not added now, the change is acceptable for me since CI is not broken.

Even if the original logic is correct (but again, [_it isn't_](https://elixir.bootlin.com/linux/v6.11.2/source/net/ipv4/tcp.c#L4054)), as a long-term solution we need to reorder the locking order of `state` and `options`. This is because `state` needs to be a mutex, but we might want to keep `options` a spin lock.

However, it's a bigger and irrelevant problem, let's leave this as future work.

**Comment by lrh2000** [comment]
Sorry for the delay in responding. I'm busy preparing for the Ph.D. preliminary exam, which will take place next week. I think I will be (almost) offline for the next few days.

**Comment by tatetian** [comment]
> Sorry for the delay in responding. I'm busy preparing for the Ph.D. preliminary exam, which will take place next week. I think I will be (almost) offline for the next few days.

Thanks for the head-up!

@StevenJiang1110 If you are mostly satisfied with this PR, then I suggest we merge it now so that we can move forward on benchmarking Ngnix.

File: kernel/src/net/socket/ip/stream/connecting.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (StevenJiang1110) [reply kernel/src/net/socket/ip/stream/connecting.rs:112]: Thanks for your explanation!

**Review by StevenJiang1110 — APPROVED** [review]
LGTM!

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
